ldap_auth_provider.py

```diff
@@ -220,6 +210,105 @@ def check_password(self, user_id, password):
             logger.warning("Error during ldap authentication: %s", e)
             defer.returnValue(False)
 
+    @defer.inlineCallbacks
+    def check_3pid_auth(self, medium, address, password):
+        """ Handle authentication against thirdparty login types, such as email
+
+            Args:
+                medium (str): Medium of the 3PID (e.g email, msisdn).
+                address (str): Address of the 3PID (e.g bob@example.com for email).
+                password (str): The provided password of the user.
+
+            Returns:
+                user_id (str|None): ID of the user if authentication
+                    successful. None otherwise.
+        """
+
+        # We currently only support email
+        if medium != "email":
+            return
+
+        # Talk to LDAP and check if this email/password combo is correct
+        try:
+            server = ldap3.Server(self.ldap_uri, get_info=None)
+            logger.debug(
+                "Attempting LDAP connection with %s",
+                self.ldap_uri
+            )
+
+            search_filter = [("mail", address)]
+            result, conn, response = yield self._ldap_authenticated_search(
+                server=server, password=password, filters=search_filter,
+                attributes=['givenname', 'cn'],
+            )
+            logger.debug(
+                'LDAP auth method authenticated search returned: '
+                '%s (conn: %s) (response: %s)',
+                result,
+                conn,
+                response
+            )
+            if not result:
+                defer.returnValue(None)
+
+            try:
+                logger.info(
+                    "User authenticated against LDAP server: %s",
+                    conn
+                )
+            except NameError:
```

Comment:
What's this doing? We don't expect logging to fail? And we've already logged conn in the debug logging above?

Reply:
Ah, this was copied from https://github.com/matrix-org/matrix-synapse-ldap3/blob/3c0b342cd0a89b33ba981b52e3178c39a2bc98d0/ldap_auth_provider.py#L136

I assume if conn logs correctly then connection was successfully established. Not sure if this is the recommended way though.

Reply:
I think the code before was checking whether we had managed to even define `conn` as a variable, rather than checking if conn can be stringified.

Reply:
I've set an option to the connection function to raise an exception if there's an issue connecting which should solve this?